### DISCUSSION TITLE
Is there a plan for homebrew-core and homebrew-cask squashing (history removal)?

### DISCUSSION THREAD

**Discussion by gibfahn** [Homebrew/discussions#5207] [Tap maintenance and Homebrew development]
### Description of issue

I know that homebrew [doesn't support shallow clones](https://github.com/orgs/Homebrew/discussions/2862) for good reason, and that the [homebrew API](https://formulae.brew.sh/docs/api/) means most users no longer have to clone the full cask and core repos, but presumably it will continue to be necessary for developers and CI for a while.

I guess my questions are really:

1. Would brew support this? I guess the key question is whether if `brew update` found a new `master` branch that didn't share history with the current one, it would be able to do the update anyway.
2. Has there been any discussion about doing this in general?

I assumed that I'd find threads on this already, but haven't found any via searching.

**Comment by Bo98** [comment]
We did plan another type of history rewrite as a part of sharding (where we split the Formula and Casks directories into subdirectories by letters). Wouldn't delete/squash history but would make it perform a bit better as new path structure is a better tree structure for git to deal with, so the plan was to make all of history use the same structure.

The big issue was this took way too long for existing clones to update. `brew update` does accept history rewrites, but a complete history rewrite made it take so long we gave up waiting. And the problem is even if we optimise the update flow further, people on older versions still matter as they may eventually update and they will always go for the latest version and skip over the intermediate optimisation.

So realistically the approach would be a completely new repository at a separate URL and give tons of notice to external update tools to support  that. There's no plans for this currently.

**Reply by gibfahn** [reply]
That makes sense as a reason why a "rewrite full git history" would be painful, I was thinking more something like:

1. Create a new root commit with the current contents of homebrew-core@master (but none of the git history)
2. force-push homebrew-core@master to a that new root commit
3. Leave the old branch in the repo as `2024-master` or something (if `brew tap homebrew/core` only fetches the `master` branch by default) or put it in https://github.com/homebrew/homebrew-core-archive

That way if people really want to see history they can add homebrew-core-archive as a remote and spelunk, but most people could just have the minimal history. Then maybe we could do that every year or something.
